source/extensions/common/aws/credential_providers/iam_roles_anywhere_x509_credentials_provider.cc

```
@@ -0,0 +1,258 @@
+#include "source/common/tls/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Aws {
+
+using std::chrono::seconds;
+
+constexpr std::chrono::hours REFRESH_INTERVAL{1};
+constexpr std::chrono::seconds REFRESH_GRACE_PERIOD{5};
+constexpr uint64_t X509_CERTIFICATE_MAX_BYTES{2048};
+
+void CachedX509CredentialsProviderBase::refreshIfNeeded() {
+  if (needsRefresh()) {
+    refresh();
+  }
+}
+
+IAMRolesAnywhereX509CredentialsProvider::IAMRolesAnywhereX509CredentialsProvider(
+    Server::Configuration::ServerFactoryContext& context,
+    envoy::config::core::v3::DataSource certificate_data_source,
+    envoy::config::core::v3::DataSource private_key_data_source,
+    absl::optional<envoy::config::core::v3::DataSource> certificate_chain_data_source)
+    : context_(context), certificate_data_source_(certificate_data_source),
+      private_key_data_source_(private_key_data_source),
+      certificate_chain_data_source_(certificate_chain_data_source) {
+
+  auto provider_or_error_ = Config::DataSource::DataSourceProvider::create(
+      certificate_data_source_, context.mainThreadDispatcher(), context.threadLocal(),
+      context.api(), false, X509_CERTIFICATE_MAX_BYTES);
+  if (provider_or_error_.ok()) {
+    certificate_data_source_provider_ = std::move(provider_or_error_.value());
+  } else {
+    ENVOY_LOG(error, "Invalid certificate data source");
+    certificate_data_source_provider_ = nullptr;
+    return;
+  }
+
+  if (certificate_chain_data_source_.has_value()) {
+    auto chain_provider_or_error_ = Config::DataSource::DataSourceProvider::create(
+        certificate_chain_data_source_.value(), context.mainThreadDispatcher(),
+        context.threadLocal(), context.api(), false, X509_CERTIFICATE_MAX_BYTES * 5);
+    if (chain_provider_or_error_.ok()) {
+      certificate_chain_data_source_provider_ = std::move(chain_provider_or_error_.value());
+    } else {
+      ENVOY_LOG(error, "Invalid certificate chain data source");
```

Same as before, it is unclear what would be the state when the config is invalid.
If possible, Envoy should reject any config that leads to a bad state, rather than silently allow it to be used.

---

As above.